src/test/java/org/jenkinsci/plugins/workflow/support/steps/build/BuildTriggerStepRestartTest.java

```diff
@@ -64,6 +78,81 @@ public void restartBetweenJobs() throws Throwable {
+            WorkflowRun upstreamRun = j.buildAndAssertSuccess(upstream);
+            // Check action while the build is still in the queue.
+            String buildStepId = BuildTriggerStepTest.findFirstNodeWithDescriptor(upstreamRun.getExecution(), BuildTriggerStep.DescriptorImpl.class).getId();
+            DownstreamBuildAction action = upstreamRun.getAction(DownstreamBuildAction.class);
+            assertThat(action.getFlowNodeId(), equalTo(buildStepId));
+            assertThat(action.getJobFullName(), equalTo(downstream.getFullName()));
+            assertThat(action.getBuildNumber(), nullValue());
+            assertThat(action.getBuild(), nullValue());
+            // Check again once the build has started.
+            j.createOnlineSlave(Label.parseExpression("agent"));
+            await().atMost(10, TimeUnit.SECONDS).until(() -> action.getBuildNumber(), notNullValue());
+            Run<?, ?> downstreamRun = downstream.getLastBuild();
+            assertThat(action.getFlowNodeId(), equalTo(buildStepId));
+            assertThat(action.getJobFullName(), equalTo(downstream.getFullName()));
+            assertThat(action.getBuildNumber(), equalTo(downstreamRun.getNumber()));
+            assertThat(action.getBuild(), equalTo(downstreamRun));
+            j.waitForCompletion(downstreamRun);
+        });
+        sessions.then(j -> {
+            FreeStyleProject downstream = j.jenkins.getItemByFullName("downstream", FreeStyleProject.class);
+            WorkflowJob upstream = j.jenkins.getItemByFullName("upstream", WorkflowJob.class);
+            WorkflowRun upstreamRun = upstream.getLastBuild();
+            String buildStepId = BuildTriggerStepTest.findFirstNodeWithDescriptor(upstreamRun.getExecution(), BuildTriggerStep.DescriptorImpl.class).getId();
+            DownstreamBuildAction action = upstreamRun.getAction(DownstreamBuildAction.class);
+            assertThat(action.getFlowNodeId(), equalTo(buildStepId));
+            assertThat(action.getJobFullName(), equalTo(downstream.getFullName()));
+            assertThat(action.getBuildNumber(), equalTo(downstream.getLastBuild().getNumber()));
+            assertThat(action.getBuild(), equalTo(downstream.getLastBuild()));
+        });
+    }
+
+    @Test
+    public void downstreamBuildActionMultipleBuilds() throws Throwable {
+        sessions.then(j -> {
+            FreeStyleProject downstream = j.createFreeStyleProject("downstream");
+            WorkflowJob upstream = j.jenkins.createProject(WorkflowJob.class, "upstream");
+            upstream.setDefinition(new CpsFlowDefinition("build(job: 'downstream'); build(job: 'downstream')", true));
+            WorkflowRun upstreamRun = j.buildAndAssertSuccess(upstream);
+            Map<String, DownstreamBuildAction> actions = upstreamRun.getActions(DownstreamBuildAction.class).stream()
+                    .collect(Collectors.toMap(a -> a.getFlowNodeId(), a -> a));
```

Comment:
I find it surprising that there are multiple `DownstreamBuildAction`s. Typically actions are singletons, with substructure as needed (compare `CauseAction.causes`).

Reply:
I was thinking of Git's `BuildData`, but sure I can refactor everything if desired.

Reply:
Do not have a strong feeling about it, though I think the singleton style is preferred as a rule. `InterruptedBuildAction.causes` is another example.

Reply:
See https://github.com/jenkinsci/pipeline-build-step-plugin/pull/128/commits/ec1d5f8a45516a234b233af384f2b299a724492d. FWIW after making the change I think if you have a case like this where the action is not created deterministically at build start (although I guess we could do that) _and_ you need to update the set of internal items dynamically using one action per internal item seems much more straightforward.